controllers/space/space_controller_test.go

```diff
@@ -0,0 +1,796 @@
+				spacetest.AssertThatSpace(t, space.Namespace, space.Name, hostClient).
+					HasStatusTargetCluster("member-1").
+					HasConditions(toolchainv1alpha1.Condition{
+						Type:    toolchainv1alpha1.ConditionReady,
+						Status:  corev1.ConditionFalse,
+						Reason:  toolchainv1alpha1.SpaceUnableToCreateNSTemplateSetReason,
+						Message: "mock error!",
+					})
+			})
+
+			t.Run("error while updating status after creating NSTemplateSet", func(t *testing.T) {
+				// given
+				space := spacetest.NewSpace("oddity", basicTier.Name,
+					spacetest.WithTargetCluster("member-1"),
+					spacetest.WithFinalizer())
+				hostClient := test.NewFakeClient(t, space, basicTier)
+				hostClient.MockStatusUpdate = func(ctx context.Context, obj client.Object, opts ...client.UpdateOption) error {
+					if _, ok := obj.(*toolchainv1alpha1.Space); ok {
+						return fmt.Errorf("mock error!")
+					}
+					return hostClient.Client.Status().Update(ctx, obj, opts...)
+				}
+				member1 := NewMemberCluster(t, "member-1", corev1.ConditionTrue)
+				member2 := NewMemberCluster(t, "member-2", corev1.ConditionTrue)
+				ctrl := newReconciler(hostClient, member1, member2)
+
+				// when
+				res, err := ctrl.Reconcile(context.TODO(), reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Namespace: space.Namespace,
+						Name:      space.Name,
+					},
+				})
+
+				// then
+				require.EqualError(t, err, "mock error!")
+				assert.Equal(t, reconcile.Result{Requeue: false}, res)
+				spacetest.AssertThatSpace(t, space.Namespace, space.Name, hostClient).
+					HasStatusTargetCluster(""). // not set
+					HasNoConditions()           // not set
+			})
+		})
+	})
+
+	t.Run("delete", func(t *testing.T) {
+
+		t.Run("success", func(t *testing.T) {
+
+			t.Run("after space was successfully provisioned", func(t *testing.T) {
+
+				t.Run("when space is deleted", func(t *testing.T) {
+					// given
+					space := spacetest.NewSpace("oddity", basicTier.Name,
+						spacetest.WithTargetCluster("member-1"),
+						spacetest.WithFinalizer(),
+						spacetest.WithDeletionTimestamp())
+					hostClient := test.NewFakeClient(t, space, basicTier)
+					nstmplSet := nstemplatetsettest.NewNSTemplateSet("oddity", nstemplatetsettest.WithReadyCondition())
+					member1Client := test.NewFakeClient(t, nstmplSet)
+					member1Client.MockDelete = func(ctx context.Context, obj client.Object, opts ...client.DeleteOption) error {
+						if nstmplSet, ok := obj.(*toolchainv1alpha1.NSTemplateSet); ok {
+							now := metav1.NewTime(time.Now())
+							nstmplSet.DeletionTimestamp = &now
+							nstmplSet.Status.Conditions = []toolchainv1alpha1.Condition{
+								nstemplatetsettest.Terminating(),
+							}
+							// instead of deleting the resource in the FakeClient,
+							// we update it with a `DeletionTimestamp`
+							return member1Client.Client.Update(ctx, obj)
+						}
+						return member1Client.Client.Delete(ctx, obj, opts...)
+					}
```

Comment:
Why do we need this? The purpose of this test is:

1. There is a Space that is being deleted
2. There is a NSTemplateSet that is ready
3. We call `Reconcile`
4. Then we should expect that the `delete` was called
5. which means that NSTemplateSet is gone
6. and Space's condition is set to terminating 
7. plus still has the Finalizer set.

If we want to verify the case when the deletion was already executed before - which means that the NSTemplateSet has DeletionTimestamp set, then we could configure the test with:
1. There is a Space that is being deleted
2. There is a NSTemplateSet that is being deleted
3. We call `Reconcile`
4. Then we should expect that the `delete` was NOT called
5. which means that NSTemplateSet is still there with the DeletionTimestap
6. and Space's condition is set to terminating 
7. plus still has still the Finalizer set.

Reply:
as I wrote before, the test mimicks the logic of k8s and the sequence of events that lead to the deletion of the Space resource, so we can cover all the cases implemented in the controller (trigger deletion of NSTemplateSet, then wait until NSTemplateSet is deleted to remove the finalizer on the Space)

Reply:
ok, but why not having one isolated unit tests per one event? As we have in all other controller unit tests?

1. one unit test that checks that the controller calls deletion (which means that it's really gone without any mocks)
2. one unit test that checks what happens when the NSTemplateSet is already being deleted
3. one unit test that checks what happens when the NSTemplateSet is not there and Space is being deleted
4. one or more unit tests for some error cases

Reply:
I don't understand: the way it is organized with subtests is exactly for this purpose!

```
t.Run("after space was successfully provisioned", func(t *testing.T) {
	t.Run("when space is deleted", func(t *testing.T) {
		t.Run("wait while NSTemplateSet is terminating", func(t *testing.T) {
			t.Run("done when NSTemplateSet is deleted", func(t *testing.T) {
				...
			})
		})
	})
})
```

Reply:
@xcoulon let me start with some general principles we all I believe follow when writing unit tests. Just for sake of making sure we are the same page :)

What we usually do in unit testes is testing a single reconcile loop:

Create a resource in the state you need it (including setting status if you expect it in the certain state)
Call the reconcile
Check that the resource is now in the state set by the controller (not by kube!). Including checking the status. And, if applicable, check that other resources created or updated by the controller.
If we need to test multiple reconcile loops (controller modifies some resources which triggers another reconcile loop) then we usually do it in separate tests (they can be encapsulated like in your example) when you re-create the state you need before the reconcile and check it after the reconcile.

And this is what you are mostly doing in this test as far as I can see. But it looks like you are skipping one step... and saving one additional test :)
If I'm not mistaken you need to test a step when the NSTemplateSet is only marked as deleted, and not actually deleted yet to make sure that the Space is still waiting for the actually deletion, correct?
If we wanted to follow the principles I described above then we would create another test for that when you manually create a NSTemplateSet with a deletion timestamp in the fake client and then reconcile the Space controller to make sure it behaves properly in this case instead of trying to emulate kube.
But... I personally think that we could cut corners sometime and do little tricks (like this) if it helps us to have tests which easier to read and maintain and covering the functionality in the same time. This is why we use some generic principles in the first place. Not because we follow the rules just because they are the rules.
If it's the case here? I don't know. I'm hesitant... I probably didn't take a good enough look at the tests to have a strong opinion :)

@xcoulon, @MatousJobanek et al WDYT?

Reply:
> @xcoulon let me start with some general principles we all I believe follow when writing unit tests. Just for sake of making sure we are the same page :)
> 
> What we usually do in unit testes is testing a single reconcile loop:
> 
> Create a resource in the state you need it (including setting status if you expect it in the certain state) Call the reconcile Check that the resource is now in the state set by the controller (not by kube!). Including checking the status. And, if applicable, check that other resources created or updated by the controller. If we need to test multiple reconcile loops (controller modifies some resources which triggers another reconcile loop) then we usually do it in separate tests (they can be encapsulated like in your example) when you re-create the state you need before the reconcile and check it after the reconcile.
> 
> And this is what you are mostly doing in this test as far as I can see. But it looks like you are skipping one step... and saving one additional test :) If I'm not mistaken you need to test a step when the NSTemplateSet is only marked as deleted, and not actually deleted yet to make sure that the Space is still waiting for the actually deletion, correct? If we wanted to follow the principles I described above then we would create another test for that when you manually create a NSTemplateSet with a deletion timestamp in the fake client and then reconcile the Space controller to make sure it behaves properly in this case instead of trying to emulate kube. But... I personally think that we could cut corners sometime and do little tricks (like this) if it helps us to have tests which easier to read and maintain and covering the functionality in the same time. This is why we use some generic principles in the first place. Not because we follow the rules just because they are the rules. If it's the case here? I don't know. I'm hesitant... I probably didn't take a good enough look at the tests to have a strong opinion :)
> 
> @xcoulon, @MatousJobanek et al WDYT?

I don't see where I "cut corners" in the series of tests here:

> If I'm not mistaken you need to test a step when the NSTemplateSet is only marked as deleted, and not actually deleted yet to make sure that the Space is still waiting for the actually deletion, correct?

this is the purpose of subtest <2> below, AFAICS

```
<1> t.Run("when space is deleted", func(t *testing.T) {                                          
<2>  t.Run("wait while NSTemplateSet is still terminating", func(t *testing.T) {
<3>    t.Run("done when NSTemplateSet is deleted", func(t *testing.T) {
```

I'm happy to discuss about the structure of the tests but frankly, if the logic is covered, should we spend more time on this topic?